src/Administration/Resources/app/administration/src/core/service/entity-mapping.service.ts

```diff
@@ -0,0 +1,132 @@
+    properties: Record<string, PropertySchema>;
+}
+
+type EntityNameMapping = Record<string, string>;
+type PropertyDefinition = Record<string, PropertySchema>;
+
+// eslint-disable-next-line sw-deprecation-rules/private-feature-declarations
+export default {
+    getEntityMapping,
+};
+
+const mappingTypesCache: Record<string, PropertyDefinition> = {};
+
+function getEntityMapping(entityName?: string, entityNameMapping?: EntityNameMapping): PropertyDefinition {
+    let schema: EntitySchema = {
+        properties: {},
+    };
+
+    if (typeof entityName === 'undefined') {
+        entityName = '';
+    }
+
+    if (entityNameMapping && Object.keys(entityNameMapping).length > 0) {
+        Object.entries(entityNameMapping).forEach(
+            ([
+                mappedKey,
+                mappedValue,
+            ]) => {
+                schema.properties[mappedKey] = {
+                    entity: mappedValue,
+                    type: 'object',
+                };
+            },
+        );
+    } else {
+        return schema.properties;
+    }
+
+    if (entityName.indexOf('.') < 1) {
+        return schema.properties;
+    }
+
+    const parts = entityName.split('.');
+    let lastEntityName = '';
+    let lastVal = lastEntityName;
+
+    parts.forEach((val) => {
+        const cleanVal = val.replace(/\[.*\]/, '');
+        const dubbedVal = val.replace(/\[.*\]/, '[0]');
+
+        if (val === '') {
+            lastEntityName = lastVal;
+        }
+
+        const property = schema.properties[cleanVal];
+
+        // Handle entity mapping
+        if (property?.entity) {
+            const entityDef = Shopware.EntityDefinition.getDefinitionRegistry().get(property.entity) as EntitySchema;
+            schema = entityDef;
+            lastEntityName = dubbedVal;
+
+            if (typeof mappingTypesCache[lastEntityName] === 'undefined') {
+                mappingTypesCache[lastEntityName] = {};
+                mappingTypesCache[lastEntityName] = handlePropertyMappings(schema.properties);
+            }
+        }
+
+        // Handle json_object type
+        if (property?.type === 'json_object') {
+            lastEntityName = dubbedVal;
+
+            if (typeof mappingTypesCache[lastEntityName] === 'undefined' && property.properties) {
+                mappingTypesCache[lastEntityName] = {};
+                mappingTypesCache[lastEntityName] = handlePropertyMappings(property.properties);
+            }
+        }
+
+        lastVal = dubbedVal;
+    });
+
+    if (lastVal === lastEntityName || !mappingTypesCache[lastEntityName]) {
+        return {};
+    }
+
+    return mappingTypesCache[lastEntityName];
+}
+
+function handlePropertyMappings(propertyDefinitions: PropertyDefinition): PropertyDefinition {
+    const blocklist: string[] = [];
+    const formatBlocklist: string[] = ['uuid'];
+
+    // Deep clone to avoid mutation
+    const clonedMapping = JSON.parse(JSON.stringify(propertyDefinitions)) as PropertyDefinition;
```

Comment:
Instead of this dirty trick to deep clone you could also use the `lodash` third party library, which is already part of our dependencies:
https://github.com/shopware/shopware/blob/c73a62bc755538034c7a2c4dced6a6d03fe7e647/src/Administration/Resources/app/administration/src/core/service/extension-api-data.service.ts#L6

specifically this function: https://lodash.info/doc/cloneDeep